snorkel/learning/utils.py

```diff
@@ -1,14 +1,86 @@
 import math
+import matplotlib
 import matplotlib.pyplot as plt
 import numpy as np
 import scipy.sparse as sparse
+import warnings
+
+from pandas import DataFrame
 
-import matplotlib
 matplotlib.use('Agg')
-import warnings
 warnings.filterwarnings("ignore", module="matplotlib")
 
 
+class Scorer(object):
+    """Abstract type for scorers"""
+    def __init__(self, test_candidates, test_labels, gold_candidate_set):
+        """
+        test_candidates: candidates corresponding to test_labels
+        test_labels: ground truth labels for test candidates
+        gold_candidate_set: full set of gold labeled candidates
+        """
+        self.test_candidates = test_candidates
+        self.test_labels     = test_labels
+        self.gold_cs         = gold_candidate_set
+
+    def score(self, test_marginals, train_marginals=None, b=0.5, set_unlabeled_as_neg=True, display=True):
+        raise NotImplementedError()
+
+
+class MentionScorer(Scorer):
+    """Scorer for mention level assessment"""
+    def score(self, test_marginals, train_marginals=None, b=0.5, set_unlabeled_as_neg=True, display=True):
+        """
+        test_marginals: array of marginals for test candidates
+        train_marginals (optional): array of marginals for training candidates
+        b: threshold for labeling
+        set_unlabeled_as_neg: set marginals at b to negative?
+        display: show calibration plots?
+        """
+        test_candidates = set()
+        test_label_array = []
+        tp = set()
+        fp = set()
+        tn = set()
+        fn = set()
+
+        for i, candidate in enumerate(self.test_candidates):
+            try:
+                test_label_index = self.test_labels.get_row_index(candidate)
+                test_label   = self.test_labels[test_label_index, 0]
+
+                # Set unlabeled examples to -1 by default
+                if test_label == 0 and set_unlabeled_as_neg:
+                    test_label = -1
```

Comment:
Looking at the code below, this `set_unlabeled_as_neg` flag isn't actually respected- it's always True.

I think we should actually just check to make sure all the test labels are in {-1,1}, else throw an error?